packages/plugins/typed-modal/src/TypedModal.js

```diff
@@ -0,0 +1,371 @@
+   *     display name as the first value and the type as the second.
+   *     Ex: [['Penguin', 'PENGUIN'], ['Giraffe', 'GIRAFFE']].
+   * @param {TypedModalMessages} opt_messages The messages for a typed modal.
+   */
+  constructor(workspace, btnCallbackName, types, opt_messages) {
+    const title = opt_messages ? opt_messages["TYPED_MODAL_TITLE"] :
+        'Create Typed Variable';
+    super(title, workspace);
+
+    /**
+     * The id of the currently selected type.
+     * @type {?string}
+     * @private
+     */
+    this.selectedType_ = null;
+
+    /**
+     * The input div holding the name of the variable.
+     * @type {HTMLInputElement}
+     * @protected
+     */
+    this.variableNameInput_ = null;
+
+    /**
+     * The div holding the list of variable types.
+     * @type {HTMLElement}
+     * @protected
+     */
+    this.variableTypesDiv_ = null;
+
+    /**
+     * The first type input.
+     * @type {HTMLInputElement}
+     * @protected
+     */
+    this.firstTypeInput_ = null;
+
+    /**
+     * The name used to register the button callback.
+     * @type {string}
+     * @private
+     */
+    this.btnCallBackName_ = btnCallbackName;
+
+    /**
+     * An array holding arrays with the name of the type and the display name
+     *     for the type. Ex: [['Penguin', 'PENGUIN'], ['Giraffe', 'GIRAFFE']].
+     * @type {Array<Array<string>>}
+     * @protected
+     */
+    this.types_ = types;
+
+    /**
+     * The messages for a typed modal.
+     * @type {!TypedModalMessages}
+     */
+    this.messages = opt_messages || {
+      "TYPED_MODAL_TITLE": "Create Typed Variable",
+      "TYPED_MODAL_VARIABLE_NAME_LABEL": "Variable Name: ",
+      "TYPED_MODAL_TYPES_LABEL": "Variable Types",
+      "TYPED_MODAL_CONFIRM_BUTTON": "Ok",
+      "TYPED_MODAL_CANCEL_BUTTON": "Cancel",
+      "TYPED_MODAL_INVALID_NAME":
+          "Name is not valid. Please choose a different name."
+    };
+
+    /**
+     * If true close the modal when the user clicks outside the modal.
+     * Otherwise, only close when user hits the 'X' button or escape.
+     * @type {boolean}
+     * @override
+     */
+    this.closeOnClick = false;
+  }
+
+  /**
+   * The messages for a typed modal.
+   * @typedef {{
+   *     TYPED_MODAL_CONFIRM_BUTTON: string,
+   *     TYPED_MODAL_VARIABLE_NAME_LABEL: string,
+   *     TYPED_MODAL_TYPES_LABEL: string,
+   *     TYPED_MODAL_CANCEL_BUTTON: string,
+   *     TYPED_MODAL_TITLE: string,
+   *     TYPED_MODAL_INVALID_NAME: string
+   * }} TypedModalMessages
+   */
+
+  /**
+   * Create a typed modal and display it on the given button name.
+   */
+  init() {
+    super.init();
+    this.workspace_.registerButtonCallback(this.btnCallBackName_, () => {
+      this.show();
+    });
+  }
+
+  /**
+   * Dispose of the typed modal.
+   * @override
+   */
+  dispose() {
+    super.dispose();
+    this.workspace_.removeButtonCallback(this.btnCallBackName_);
+    this.variableNameInput_ = null;
```

Comment:
Explicitly setting these to null is an old blockly pattern, but I don't think you need it for these--the garbage collector will get them--and we're phasing it out in blockly more generally.

Reply:
done